### ISSUE TITLE
Optimize BaseQuantizer Forward Pass

### ISSUE THREAD

**Issue by Shehrozkashif** [openvinotoolkit/nncf#3898]
## Description

The `BaseQuantizer._forward_impl` in the Torch backend contains several conditional checks that execute on every forward pass. These include debug counters, enablement checks, and tracing state checks.

While each condition is inexpensive individually, the quantizer runs in the hot path of deep models and is invoked extremely frequently. The cumulative overhead becomes measurable in large workloads.

This results in unnecessary branching during the forward pass and degrades performance.

---

## Location

```
layers.py:L432–451
```

---

## Problem Summary

The current implementation repeatedly evaluates internal state flags during execution:

- quantizer enablement
- debug counters
- tracing mode
- runtime state checks

These decisions are static most of the time and do not need to be recomputed on every forward call.

---

## Suggested Fix

Refactor the quantizer to use a **state-based dispatch approach**:

- Maintain separate forward implementations for each state
- Switch the forward pointer when the quantizer state changes
- Eliminate repeated branching inside the hot path

Example idea:

```python
self.forward = self._forward_enabled
# or
self.forward = self._forward_disabled
```

This moves decision-making outside the hot loop and keeps execution minimal.

---

## Expected Benefits

- Reduced branching in the forward path
- Lower runtime overhead in deep models
- Cleaner separation of quantizer states
- Improved performance in training and inference

**Comment by AlexanderDokuchaev** [comment]
Hi @Shehrozkashif 
Could you please provide performance metrics on how much these checks slow down the forward pass?

**Comment by newbhumika** [comment]
Hi @AlexanderDokuchaev,

I’d like to take ownership of this optimization.

Since BaseQuantizer._forward_impl sits directly in the hot path, even minimal branching overhead can accumulate significantly in deep models where quantizers are executed thousands of times per iteration. Moving state decisions outside the forward loop is a well-established performance pattern and should produce measurable gains.

I will:

Benchmark the current implementation using torch.profiler

Provide microbenchmarks isolating _forward_impl

Measure end-to-end impact on representative models

Implement a state-based dispatch approach (switching self.forward on state change)

Ensure no regression in tracing/debug behavior

Share detailed performance comparisons in the PR

I’ll proceed with benchmarking and submit results along with the proposed refactor.

**Comment by Shehrozkashif** [comment]
Hi @AlexanderDokuchaev,

I have conducted a **high-precision benchmarking analysis** using the real  
`SymmetricQuantizer` implementation to quantify the impact of these checks.

---

## Performance Metrics (Actual NNCF Codebase)

The following metrics were captured by running `BaseQuantizer._forward_impl` through **500,000 iterations** in a controlled environment to average out system noise.

| Metric | Measured Value |
|------|---------------|
| Current Forward Implementation | ~107.12 µs per call |
| Optimized Forward (Mechanism only) | ~92.56 µs per call |
| Measured Overhead per Quantizer | ~14.56 µs (14,560 ns) |
| Mechanism Latency Reduction | ~13.6% |

---

## Analysis of the Bottleneck

While individual `if/else` branches are inexpensive, the **cumulative overhead** is significant due to the following factors:

- **Expensive State Access**  
  The current enablement check calls `self.enabled[0].item()`.  
  In PyTorch, `.item()` is an expensive cross-boundary operation that synchronizes the CPU and pulls data from the internal C++ tensor core into the Python interpreter.  
  This is being repeated on every forward pass, even though the state is static during inference.

- **Cumulative Latency**  
  In deep models with a large number of quantizers (e.g., **~2,000 layers**), this redundant logic adds **~29 ms of avoidable latency** to every single forward pass.

- **Hot Path Efficiency**  
  Moving these decisions to a **state-based dispatch** (swapping the `forward` pointer only when enable/disable is called) reduces the mechanism overhead to nearly zero, leaving only the core quantization math in the hot path.

**Comment by AlexanderDokuchaev** [comment]
@Shehrozkashif Thank you for the analysis, but this is not a critical overhead for large models.
The root cause is that the quantizer contains outdated functionality, it will be resolved later during the refactoring of the quantizers, but optimization of pass only small part of it. 

@newbhumika thanks for interesting. If you have some simple and fast solution you are free to create PR. But if this requires some time for it, then it’s better to look on a another issue.
